Issue #761: Links fail for extracting subtrees

enkore opened the pull request:
**Issue summary:**

When extracting subtrees ("partial_extract") by using any of (excludes, PATH, strip-components) hard links might fail, because Borg tried to link to non-existing files, if the first appearance of a (inode, device) tuple (which uniquely identifies files and is identical across hard links) was not extracted.

**Implementation:**

* creating a mapping that stores *all* file paths so far and their chunk list or link source. (We need to store all seen files, because we can't tell if a file is hard-linked)
* when extracting a hard link: resolving whether an instance of the link was extracted
  * if it was not, the file contents will be extracted
     * and the entry in our "master file mapping" will be changed to point to this file
  * if it was, a hard link is created

**Drawbacks:**

* a possibly very significant increase in memory usage when extracting subtrees, because the "master file mapping" will essentially grow until all files in the archive were visited, and at that point it will contain all file paths and chunk lists.

However, if no partial extraction is done, the "MFM" is not created, not used and has therefore no impact.

**Won't somebody please think of the ~~children~~ soft links?**

Soft links are always stored as absolute paths and always extracted as such and are not affected by this.

**TODO**

* [x] check whether this handles link metadata correctly / as expected, create test or validate that test exists and indeed covers this additional code path
   * they are the same files, so same U/G ID, permissions, xattrs and ACLs, even on Windows/NTFS.
* [x] Xcode builds broken (some minor thing with output paths or something)
* [x] This "breaks" preloading (now only for "old" archives)
* [x] Memory hit (now only for "old" archives)
   * [x] DONE We could include a nlink / a flag whether nlink > 1 for hard link items. That way the "MFM" would only need to contain these items, and items can be directly filtered again (patternmatcher or nlink), which also fixes preloading (*but only for new archives*). We can't use this for reverse ref-counting, though, because nlink = x doesn't necessarily imply that there are (x-1) items hardlinked to it.

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -298,7 +298,36 @@ def add_file_chunks(chunks):
+        Extract archive item.
+
+        :param item: the item to extract
+        :param restore_attrs: restor file attributes
```
typo restore

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -298,7 +298,36 @@ def add_file_chunks(chunks):
+        :param restore_attrs: restor file attributes
+        :param dry_run: do not write any data
+        :param stdout: write extracted data to stdout
+        :param sparse: recreate sparse files (only approximate, no exact hole reproduction)
```
in fact it is even a "sparsify" of all files that have all-zero chunks, no matter whether they were holes a archive creation time or not.

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -327,25 +356,36 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
+                if make_link:
+                    os.link(link_target, path)
+                    return
+                # we will extract the item "by chunks", since the items providing the chunks has not been extracted
```
typo item

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -327,25 +356,36 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
+                    os.link(link_target, path)
+                    return
+                # we will extract the item "by chunks", since the items providing the chunks has not been extracted
+                # hence we overwrite it's entry with out entry, so that future hardlinks will correctly link
```
typo its, typo our

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -327,25 +356,36 @@ def extract_item(self, item, restore_attrs=True, dry_run=False, stdout=False, sp
-                    fd.truncate(pos)
-                    fd.flush()
-                    self.restore_attrs(path, item, fd=fd.fileno())
+                if not items:
```
so, this "if" would trigger if items is None and also if items == {} - is this wanted?

enkore replied:
If items were empty and we're extracting a hard link from a subtree, then archive is corrupted, since the item holding the chunks for the hard link didn't come first.

ThomasWaldmann replied:
ok, so the code should be "if items is None" as the other case is not possible?
maybe also check "if not items" afterwards and raise an error?

enkore replied:
If this were somehow to happen (and I can't see how, assuming that the archive is as-created), then this will blow up and throw an exception anyway when trying to look the b'source' up in it.

enkore commented on borg/archive.py:
```diff
@@ -298,35 +298,16 @@ def add_file_chunks(chunks):
         cache.rollback()
```
"This over avoids storing all files in a mapping and allows proper preloading for new archives."

Oops. Guess I over-pasted part of the commit message :D

ThomasWaldmann commented on borg/archiver.py:
```diff
@@ -355,9 +355,10 @@ def do_extract(self, args):
+                filter=lambda item: item.get(b'link', True) or matcher.match(item[b'path'])):
             orig_path = item[b'path']
-            if partial_extract and stat.S_ISREG(item[b'mode']):
+            if partial_extract and stat.S_ISREG(item[b'mode']) and item.get(b'link', True):
```
so for old archives, it will assume all files could be hardlinks (either with data or referring to another hardlink).

enkore replied:
Yes, with all the drawbacks mentioned above (a lot higher memory usage, broken/useless preloading for remote repositories, since all items are marked for preloading as if you were extracting the entire archive).

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -589,7 +572,10 @@ def process_file(self, path, st, cache, ignore_inode=False):
-        item = {b'path': safe_path}
+        item = {
+            b'path': safe_path,
+            b'link': st.st_nlink > 1,
```
so the meaning of "link" is that this item is member of a set of hardlinks (the first one will have data, the other members will refer to the first one). maybe this should be in a comment in the src.

ThomasWaldmann replied:
hm, no. meaning of "link" is rather "this item has the data for a set of hardlinks".

enkore replied:
Not exactly, it means that this is item is a hard link (nlink > 1) and that it has the chunks. That way the "items" mapping only holds items which are actually needed to recreate a hard link.

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -574,7 +574,7 @@ def process_file(self, path, st, cache, ignore_inode=False):
         item = {
             b'path': safe_path,
-            b'link': st.st_nlink > 1,
+            b'link': st.st_nlink > 1,  # item is a hard link and has the chunks
```
considering that every regular file is somehow a "hard link", the entry name could just be "multilinks".

enkore replied:
How about b'linkmaster' or b'mainlink'? Makes it clear that this about a link and it's somehow important / different...

ThomasWaldmann replied:
hardlink_master plus a note that we consider the first hardlink we encounter as master and save it with the data.

ThomasWaldmann commented on borg/archive.py:
```diff
@@ -549,7 +549,7 @@ def process_file(self, path, st, cache, ignore_inode=False):
                     b'path': safe_path,
                     b'source': source,
-                    b'link': False,
+                    b'link': False,  # item is a hard link (and doesn't have the chunks)
```
hmm, do we need it here? we have already source.

enkore replied:
No we don't, items with b'source' can never be in `items` (in do_extract).

ThomasWaldmann commented:
btw, if a commit comment ends with ", fixes #XXX", github will automatically close the issue with a link to the changeset that closed it.

enkore commented:
I added another sub-testcase, which wasn't handled well previously (when the hardlink_master was extracted as well and this special-case-code was effective, then it would be extracted twice, even though a hard link to the master was possible).

I think this should cover it. The xcode tests should pass now, their failure were related to the absolute paths of some symlinks, I just removed that code, had nothing to do with this anyway.

The open question is metadata handling, I'll look into that tomorrow.

ThomasWaldmann commented on borg/archiver.py:
```diff
@@ -353,8 +353,20 @@ def do_extract(self, args):
+        hardlink_masters = {} if partial_extract else None
+
+        def item_is_hardlink_master(item):
+            return (partial_extract and stat.S_ISREG(item[b'mode']) and
```
nice name, but not totally true. if partial_extract is False, it returns False even if it is a hardlink_master.

ThomasWaldmann replied:
so, just drop function and inline it?

enkore replied:
It has two call sites, didn't want to duplicate the code. (one in the filter lambda, the other in the if below to only add possible masters to hardlink_masters).

ThomasWaldmann replied:
ah, ok.

dyu commented:
I still get an error if a hardlink is extracted ```[Errno 2] No such file or directory```

Here's a test script to reproduce
```sh
#!/bin/sh

mkdir test && cd test

#initial data
mkdir data
echo "a" > data/a.txt

#initial backup
mkdir backups && cd backups
BORG_PASSPHRASE=bar! borg init bar
BORG_PASSPHRASE=bar! borg create bar::b1 ../data

# update
echo "bb" > ../data/b.txt
BORG_PASSPHRASE=bar! borg create bar::b2 ../data

#extract
mkdir extracted && cd extracted
BORG_PASSPHRASE=bar! borg extract -v --list ../bar::b2
cd ..

cd ../data
# hardlink
ln b.txt c.txt

cd - > /dev/null
BORG_PASSPHRASE=bar! borg create bar::b3 ../data

mkdir extracted-hardlink && cd extracted-hardlink
# this fails (works fine if data/b.txt is provided as arg)
BORG_PASSPHRASE=bar! borg extract -v --list ../bar::b3 data/c.txt


```

enkore commented:
Version? This fix has not been released yet.
